File: solutions/java/say/src/main/java/Say.java
public final class Say {
    public static final long MAXIMUM_PRONOUNCEABLE_NUMBER = 999_999_999_999L;
    public static final long HUNDRED = 100;
    public static final long THOUSAND = 1_000;
    public static final long MILLION = 1_000_000;
    public static final long BILLION = 1_000_000_000;

    public String say(long number) {
        validateNumber(number);
        return sayNumber(number);
    }

    private void validateNumber(long number) {
        if (number < 0 || number > MAXIMUM_PRONOUNCEABLE_NUMBER) {
            throw new IllegalArgumentException("Number must be between 0 and " + MAXIMUM_PRONOUNCEABLE_NUMBER + ".");
        }
    }

    public String sayNumber(long number) {
        return switch ((int) number) {
            case 0 -> "zero";
            case 1 -> "one";
            case 2 -> "two";
            case 3 -> "three";
            case 4 -> "four";
            case 5 -> "five";
            case 6 -> "six";
            case 7 -> "seven";
            case 8 -> "eight";
            case 9 -> "nine";
            case 10 -> "ten";
            case 11 -> "eleven";
            case 12 -> "twelve";
            case 13 -> "thirteen";
            case 14 -> "fourteen";
            case 15 -> "fifteen";
            case 16 -> "sixteen";
            case 17 -> "seventeen";
            case 18 -> "eighteen";
            case 19 -> "nineteen";
            case 20 -> "twenty";
            case 30 -> "thirty";
            case 40 -> "forty";
            case 50 -> "fifty";
            case 60 -> "sixty";
            case 70 -> "seventy";
            case 80 -> "eighty";
            case 90 -> "ninety";
            default -> {
                if (number < HUNDRED) yield sayNumber(number / 10 * 10) + "-" + sayNumber(number % 10);
                if (number < THOUSAND) yield sayMagnitude(number, HUNDRED, "hundred");
                if (number < MILLION) yield sayMagnitude(number, THOUSAND, "thousand");
                if (number < BILLION) yield sayMagnitude(number, MILLION, "million");
                yield sayMagnitude(number, BILLION, "billion");
            }
        };
    }

    private String sayMagnitude(long number, long divisor, String name) {
        var quotientPart = sayNumber(number / divisor) + " " + name;
        var remainder = number % divisor;
        return remainder == 0 ? quotientPart : quotientPart + " " + sayNumber(remainder);
    }
}

File: solutions/scala/say/src/main/scala/Say.scala
object Say:
  private val MaxPronounceableNumber = 999_999_999_999L
  private val Billion = 1_000_000_000L
  private val Million = 1_000_000
  private val Thousand = 1_000

  def inEnglish(number: Long): Option[String] =
    Option(sayNumber(number))

  private def sayNumber(number: Long): String =
    def sayMagnitude(divisor: Long, name: String) =
      val quotientPart = sayNumber(number / divisor) + " " + name
      number % divisor match
        case 0 => quotientPart
        case remainder => quotientPart + " " + sayNumber(remainder)

    number match
      case x if x < 0 || MaxPronounceableNumber < x => null
      case 0 => "zero"
      case 1 => "one"
      case 2 => "two"
      case 3 => "three"
      case 4 => "four"
      case 5 => "five"
      case 6 => "six"
      case 7 => "seven"
      case 8 => "eight"
      case 9 => "nine"
      case 10 => "ten"
      case 11 => "eleven"
      case 12 => "twelve"
      case 13 => "thirteen"
      case 14 => "fourteen"
      case 15 => "fifteen"
      case 16 => "sixteen"
      case 17 => "seventeen"
      case 18 => "eighteen"
      case 19 => "nineteen"
      case 20 => "twenty"
      case 30 => "thirty"
      case 40 => "forty"
      case 50 => "fifty"
      case 60 => "sixty"
      case 70 => "seventy"
      case 80 => "eighty"
      case 90 => "ninety"
      case n if n < 100 => sayNumber(n / 10 * 10) + "-" + sayNumber(n % 10)
      case n if n < Thousand => sayMagnitude(100, "hundred")
      case n if n < Million => sayMagnitude(Thousand, "thousand")
      case n if n < Billion => sayMagnitude(Million, "million")
      case _ => sayMagnitude(Billion, "billion")
  end sayNumber
end Say